### DISCUSSION TITLE
invoking --help when using quarkus with picocli and a camel quarkus extension shows also camel quarkus help (not only picocli)

### DISCUSSION THREAD

**Discussion by guykatz** [quarkusio/quarkus#34793] [Q&A]
Hi guys;
I wasnt sure where to put this question as it has many parts... but I like this community :)
I am using quarkus to build a command-line application using picocli.
my application uses also gson and jsonpath libs. I found both libs in the camel quarkus extensions module and figured its better to use these extensions rather than just adding them as dependencies in the build file.
everything until now works nice functionally.

the only thing I dont like visually is when invoking the application. it will show first the camel quarkus help (which is long) and only then is will show the actual quarkus picocli app output (which Is the only thing I want to see). moreover, when invoking the picocli quarkus app with the needed commands, camel quarkus will show its help first, along with a warning that  the command is unknown but the app will then launch properly and output the proper data
I guess my questions are:
1. should I even use camel quarkus extensions over just adding straight dependency? does it really promise native compilcation ability this way?
2. is there a way to disable the camel quarkus logging here?
3. why is camel a 'player' here. why does it think I am invoking it at all. its a picocli app...
your help is appreciated
here is an output when app is used with proper commands:

> PS C:\dev_env\quarkus-picocli> java -jar build/quarkus-app/quarkus-run.jar -c=get -p="somepath"
> Unknown option: -c=get -p=somepath
> 
> Apache Camel Runner takes the following options
> 
>   -h or -help = Displays the help screen
>   -r or -routers <routerBuilderClasses> = Sets the router builder classes which will be loaded while starting the camel context
>   -d or -duration <duration> = Sets the time duration (seconds) that the application will run for before terminating.
>   -dm or -durationMaxMessages <durationMaxMessages> = Sets the duration of maximum number of messages that the application will process before terminating.
>   -di or -durationIdle <durationIdle> = Sets the idle time duration (seconds) duration that the application can be idle before terminating.
>   -t or -trace = Enables tracing
>   -ts or -traceStandby = Enables tracing standby
>   -e or -exitcode <exitcode> = Sets the exit code if duration was hit
>   -pl or -propertiesLocation <propertiesLocation> = Sets location(s) to load properties, such as from classpath or file system.
> *************** [55]

the last line with the ***[55] is actually the only line i expected , not the whole camel thing.

here is the output when invoking help:

> PS C:\dev_env\quarkus-picocli> java -jar build/quarkus-app/quarkus-run.jar -h
> Apache Camel Runner takes the following options
> 
>   -h or -help = Displays the help screen
>   -r or -routers <routerBuilderClasses> = Sets the router builder classes which will be loaded while starting the camel context
>   -d or -duration <duration> = Sets the time duration (seconds) that the application will run for before terminating.
>   -dm or -durationMaxMessages <durationMaxMessages> = Sets the duration of maximum number of messages that the application will process before terminating.
>   -di or -durationIdle <durationIdle> = Sets the idle time duration (seconds) duration that the application can be idle before terminating.
>   -t or -trace = Enables tracing
>   -ts or -traceStandby = Enables tracing standby
>   -e or -exitcode <exitcode> = Sets the exit code if duration was hit
>   -pl or -propertiesLocation <propertiesLocation> = Sets location(s) to load properties, such as from classpath or file system.
> Usage: greeting [-hV] [-c=<command>] [-p=<jsonPath>] <name>
>       <name>                Your name.
>   -c, --command=<command>   Command to use. either get ot set
>   -h, --help                Show this help message and exit.
>   -p, --path=<jsonPath>     the json path to apply to the object graph. use $
>                               for all tree
>   -V, --version             Print version information and exit.

[a comment by an automated account is not shown]

**Comment by ebullient** [comment]
I have a feeling you've fallen into (or are not avoiding) the setup/creation of Camel's default handler. Remember that Quarkus extensions do more than just configure what you'd need to run things natively: extensions also do a lot of auto-configuration of beans, etc.

Which extensions do you have enabled?  A subset of your project (enough to illustrate the problem), would really help.

In the meanwhile, I have a PicoCLI app that also uses Camel here: https://github.com/ebullient/pockets-cli. I'm using Camel directly, but only sometimes, so I've done some things explicitly to control when and how Camel starts, so you may be able to find what you need by poking around in there.

**Reply by guykatz** [reply]
thanks @ebullient 
the "issue" is also reproduced on a barebones project with the following extensions: picocli, camel jsonpath, camel gson.
all u need is to build and then launch the app using 'java -jar build/quarkus-app/quarkus-run.jar erin'
you will get the camel help for not undetanding the 'erin' parameter right away and then the app will continue as usual and output the correct data.
also, took a look at your app and tried to use some of the camel related properties but my issue still persists...
thanks

**Comment by ebullient** [comment]
If you haven't provided an alternate main, the default Camel main will register itself:

```
% java -jar target/quarkus-app/quarkus-run.jar 
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2023-07-19 10:13:06,462 INFO  [org.apa.cam.qua.cor.CamelBootstrapRecorder] (main) Bootstrap runtime: org.apache.camel.quarkus.main.CamelMainRuntime
2023-07-19 10:13:06,484 INFO  [org.apa.cam.mai.MainSupport] (main) Apache Camel (Main) 4.0.0-M3 is starting
2023-07-19 10:13:06,513 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Apache Camel 4.0.0-M3 (camel-1) is starting
2023-07-19 10:13:06,514 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Routes startup (started:0)
2023-07-19 10:13:06,514 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Apache Camel 4.0.0-M3 (camel-1) started in 1ms (build:0ms init:0ms start:1ms)
2023-07-19 10:13:06,517 INFO  [io.quarkus] (main) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 3.2.0.Final) started in 0.540s. 
2023-07-19 10:13:06,517 INFO  [io.quarkus] (main) Profile prod activated. 
2023-07-19 10:13:06,517 INFO  [io.quarkus] (main) Installed features: [camel-core, camel-gson, camel-jsonpath, cdi, picocli]
Hello picocli, go go commando!
2023-07-19 10:13:06,569 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Apache Camel 4.0.0-M3 (camel-1) is shutting down (timeout:45s)
2023-07-19 10:13:06,571 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Routes stopped (stopped:0)
2023-07-19 10:13:06,572 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Apache Camel 4.0.0-M3 (camel-1) shutdown in 3ms (uptime:0s)
2023-07-19 10:13:06,577 INFO  [io.quarkus] (main) code-with-quarkus stopped in 0.008s
```

Note the enabled extensions (based on dependencies):
```
Installed features: [camel-core, camel-gson, camel-jsonpath, cdi, picocli]
```

You need to provide a main class for your CLI that uses the `@QuarkusMain` annotation to ensure it takes precedence over the default provided by Camel. You should also disable camel's default bootstrap behavior using [quarkus.camel.bootstrap.enabled](https://camel.apache.org/camel-quarkus/3.0.x/reference/extensions/core.html#quarkus.camel.bootstrap.enabled).

In pockets, I set the following (to slim down startup activity):

```
quarkus.camel.bootstrap.enabled=false
quarkus.camel.runtime-catalog.components=false
quarkus.camel.runtime-catalog.languages=false
quarkus.camel.runtime-catalog.dataformats=false
quarkus.camel.runtime-catalog.models=false
```
